File: 2015/06-Fire-Hazard.hs
import Data.List.Split
import Data.Matrix

import qualified Data.Set as S
import Utils

data Operation = ON | OFF | TOGGLE deriving (Eq, Show)
type Point = (Int, Int)
type Instruction = (Operation, Point, Point)

parseInput :: [String] -> [Instruction]
parseInput = map (parseLine.splitOn " ")
 where
   parseLine ("turn":"on":xs)   = parseNums ON xs
   parseLine ("turn":"off":xs)  = parseNums OFF xs
   parseLine ("toggle":xs)      = parseNums TOGGLE xs
   parseLine _                  = error "Unknown operation"

   parseNums op [a,"through",b] = (op, point a, point b)
   parseNums _ _                = error "unknown interval"

   point = tuple . atoiList'

   tuple [a,b] = (a, b)
   tuple _     = error "not a tuple"

points :: Instruction -> (Operation, [Point])
points (op, (p1, p2), (q1, q2)) = (op, [(p,q) | p <- [p1..q1], q <- [p2..q2]])

complement :: Ord a => S.Set a -> S.Set a -> S.Set a
complement a b = S.difference (S.union a b) (S.intersection a b)

partOne instructions = length $ foldl switch S.empty $ map points instructions
  where
    switch acc (ON, p) = S.union acc (S.fromList p)
    switch acc (OFF, p) = S.difference acc (S.fromList p)
    switch acc (TOGGLE, p) = complement acc (S.fromList p)


partTwo instructions = foldl switch (zero 1000 1000) $ map points instructions
  where
    switch m (ON, p) = foldl (incElem 1) m p
    switch acc (OFF, p) = foldl decElem acc p
    switch acc (TOGGLE, p) = foldl (incElem 2) acc p
    incElem n m coord = setElem ((m ! coord) + n)  coord m
    decElem m coord = setElem (max ((m ! coord) - 1) 0) coord m


main = do
  contents <- getContents
  let input = lines contents
      instructions = parseInput input

  print $ partOne instructions -- 400410

  print $ partTwo instructions

File: 2015/06-Fire-Hazard.js
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin })

const instructions = []

const ints = str => str.split(',').map(Number)

rl.on("line", data => {
  const [operation, from, to] = data.replace("turn ", "").replace("through ", "").split(" ")


  instructions.push([operation, ints(from), ints(to)])
})


const englishOperations = {
  on: x => 1,
  off: x => 0,
  toggle: x => x === 1 ? 0 : 1
}

const elfOperations = {
  on: x => x+1,
  off: x => Math.max(0, x-1),
  toggle: x => x + 2
}

function * coordinates ([x1,y1], [x2,y2]) {
  for (let x=x1; x<=x2; x+=1)
  for (let y=y1; y<=y2; y+=1)
  yield [x,y]
}

function solve (operations) {
  const lights = new Uint8Array(1000*1000)

  for (const [op, from, to] of instructions)
  for (let [x,y] of coordinates(from, to)) {
    const pos = 1000*x + y
    lights[pos] = operations[op](lights[pos])
  }
  return lights.reduce((a,b)=>a+b)
}

rl.on('close', () => {
  console.log('partOne', solve(englishOperations))
  console.log('partTwo', solve(elfOperations))
})
